README.md
# showcase
Personal project portfolio and independent learning showcase

**showcase/dissertation_program:** My python program to code participant's responses on 5 psychometric scales. See references in corresponding python module under showcase/dissertation_program/scales/<scale_name>.py
1. Creative Achievement Questionnaire
2. Creative Self-efficacy
3. Alternative Uses and Instances
4. International Personality Item Pool
5. Hagen's Matrices Test (Short-form version)
- **(Detailed Description)** The purpose of this dissertation program is to take the raw data (e.g.: diss_survey_22 October 2022_19.48.csv) which contains participants' uncalculated answers for psychometric scales, calculate their scores and export them in a format which is friendly for further analysis in IBM SPSS Statistics (e.g.: showcase/dissertation_program/scales_profiles/output_ipip50.csv).


**showcase/codewars:** My Codewars practise example showcase

**showcase/codewars/data_science:** Codewars practise examples that are related to data science
- Any folder beginning 'ex' contains a solution module and a unittest module which begins with 'test_'


dissertation_program/changelog.txt
-=-=-=-=-=-To-Do-=-=-=-=-=-
- Nothing!?


### Impractical Possibilities ###
# [] Add way to manually input qualtrics column names for each module
# [] Data handling between-subjects and within-subjects
- Each scale/module will need to create its own table 4/5 completed [auaifluency,cse,hmts,ipip50]
--> Recode create tables and writing tables to account for 'null' entries
---> could be simply? Or not? Custom SQL code to be written to not submit an entry where field == null
----> would have to supply one less parameter ot the SQL query.
- consider making scales inherit methods for checking for missing entries

## Notes ##
---* i will need to check fluency manuaally because people can put in typos


-=-=-=-=Change-log-=-=-=-=-

A-long-time-ago
- created function for measing participant completion times on qualtrics questions
- created function to parse qualtrics data and print it in a friendly/pretty format
- created participant function
16-10-22
- created class for scales
- created method for registering scales
- created method framework for creating scales' SQL tables (createtables)
- created method framework for creating scales' csv output (preparecsv) for further SPSS analysis
- made participant profiles object-oriented
- integrated scales' data with participant profiles (setprofile)
- created method to compile csv output from multiple scales and participants into single csv file
- added preparecsv for auaifluency,cse,hmts
- added createtables for auaifluency,cse,
17-10-22
- added hmts createtables
- added ipip50 createtables
- added ipip50 preparecsv
- added caq preparecsv
- edited writecsv registerparticipants module function
- timming pid off of concated csvs from multiple scales
20-10-22
- done CSE (Clean missing data with 'null' insertions)
- done hmts
21-10-22
- done IPIP50
- done auaifluency
- done CAQ
-* create contingencies for missing entries
-* account for making reponses null if they only contain whitespace with nothign else
1 - done - add auto-incremented case numbers so i can differentiate between datasets
2 - done - I need to capture the demographic data - answers to other questions and include them in the csv that i will be working from.
3 - DEMO GRAPHICS i might want to also add date of submission to my csv data collection - so if i change ethics i can find which data has given consent
22-10-22
- fixed auaifluency multiple commas as multiple auaifluency answers
- added NULL_FLAGS to replace 'null'
- Created NULL_FLAGS for Empty scale data versus Incomplete scale data Done: HMTS, CSE, auaifluency, ipip50, CAQ
- CAQ ***Missing Data (skipped) domains is being treated as 0 - See line ~163+*** 
24-10-22
- Test this program's automatic coding against manual coding of data.


-=-=-=-=Thought-log-=-=-=-=-
- trim off pid because there will be multiple columns of it from each of the scales
- I can do this by removing header[0] and row[0] after the first scale
- Each scale/module will need to prepare its own csv 5/5 completed
- [auaifluency,cse,hmts,ipip50,caq]
- Problem: scores 0 i need to input NULL values instead
-- To deal with missing cases i need to gather the field names of the scales' questions and put them into a table.
-- Then I can query across each response and fill the values with NULL
-- I should have the program ask for a model response ID, this ID will be a manually completed response with all of the questions answered
19-10-22
This is not needed for simply computing data. This is data aggregation. Another project. [- what if instead of single tables for each module, we created a single table to comprise all the data?
-- alternative to this, we need to include a key of the study ID so that cross-comparisons can be made
- I need to know what Qualtrics data, multiple scales and tables would look like in pythonanywhere admin OR single tables AND how this would be sorted]
20-10-22
- consider making scales inherit methods for checking for missing entries
-- contingencies will need to be addressed in each individual module
-- OR in fact afterwards, in clean up, comparing each participants entries... ill need to do some thinking
--- to do this i can create a question template for each qualtrics scale then compare keys to find the missing entries?
22-10-22
- # (caq-model participant template creation) This makes me think though if this is creating templates, does it create templates when the times_i/ii questions are chosen?
- # -> yes because we are only check that the main domain question has been answered. the times_i/ii scores are dealt with sepearately and if answered and empty it becomes 1


dissertation_program/scales/ipip50.py
# scale: International Personality Item Pool
# scale source: Goldberg, L. R., Johnson, J. A., Eber, H. W., Hogan, R., Ashton, M. C., Cloninger, C. R., & Gough, H. C. (2006). The International Personality Item Pool and the future of public-domain personality measures. Journal of Research in Personality, 40, 84-96. 

import re
import registerscales
import registerparticipants

class IPIP50:
    def __init__(self):
        self.name = 'ipip50'
        self.score = {
            'openness': 0,
            'conscientiousness': 0,
            'extraversion': 0,
            'agreeableness': 0,
            'emotional_stability': 0
        }
        self.questionprefix = 'ipip_fifty_'
        self.ocean = {
            'extraversion' : { #1 -- see administering ipip
                '1': 'positive',
                '11' : 'positive',
                '21' : 'positive',
                '31' : 'positive',
                '41' : 'positive',
                '6' : 'negative',
                '16' : 'negative',
                '26' : 'negative',
                '36' : 'negative',
                '46' : 'negative',
            },
            'agreeableness' : { #2
                '2': 'negative',
                '12' : 'negative',
                '22' : 'negative',
                '32' : 'negative',
                '7' : 'positive',
                '17' : 'positive',
                '27' : 'positive',
                '37' : 'positive',
                '42' : 'positive',
                '47' : 'positive',
            },
            'conscientiousness' : { #3
                '3': 'positive',
                '13' : 'positive',
                '23' : 'positive',
                '33' : 'positive',
                '43' : 'positive',
                '48' : 'positive',
                '8' : 'negative',
                '18' : 'negative',
                '28' : 'negative',
                '38' : 'negative',
            },
            'emotional_stability' : { #4
                '4': 'negative',
                '9' : 'positive',
                '19' : 'positive',
                '14' : 'negative',
                '24' : 'negative',
                '29' : 'negative',
                '34' : 'negative',
                '39' : 'negative',
                '44' : 'negative',
                '49' : 'negative',
            },
            'openness' : { #5
                '5': 'positive',
                '10' : 'negative',
                '15' : 'positive',
                '20' : 'negative',
                '25' : 'positive',
                '30' : 'negative',
                '35' : 'positive',
                '40' : 'positive',
                '45' : 'positive',
                '50' : 'positive',
            }
        }
        self.positivekeys = {
            'very inaccurate': 1,
            'moderately inaccurate': 2,
            'neither inaccurate nor accurate': 3,
            'neither accurate nor inaccurate': 3,
            'moderately accurate': 4,
            'very accurate': 5
        }
        self.negativekeys = {
            'very inaccurate': 5,
            'moderately inaccurate': 4,
            'neither inaccurate nor accurate': 3,
            'neither accurate nor inaccurate': 3,
            'moderately accurate': 2,
            'very accurate': 1
        }

        self.dimensionranges = dict()

        for dimension,dic in self.ocean.items():
            self.dimensionranges[dimension] = {}

            for qnum,sign in dic.items():
                # i need to map the question to its value and sign
                # then for all the negative questions, total them
                # then for all the positive questions, total them
                try:
                    self.dimensionranges[dimension][sign]['count'] += 1
                except:
                    self.dimensionranges[dimension][sign] = {'count' : 1,}
        
        for dimension in self.dimensionranges:
            for sign in self.dimensionranges[dimension]:
                i=1
                while i <= self.dimensionranges[dimension][sign]['count']:
                    try:
                        self.dimensionranges[dimension][sign]['total'] += 5 * (sign == 'negative' and -1 or 1)
                    except:
                        self.dimensionranges[dimension][sign]['total'] = 5 * (sign == 'negative' and -1 or 1)
                    i += 1
                
            self.dimensionranges[dimension]['median'] = (self.dimensionranges[dimension]['negative']['total'] + self.dimensionranges[dimension]['positive']['total']) / 2

    # def addcolumn(self,colname):
    #     if colname not in registerscales.DPmScales().getscalecolumns(self.name):
    #         registerscales.DPmScales().registerscalecolumn(self.name,colname)

    def calculate(self,responsedict):
        # Ensures all questions for a dimension are answered otherwise dimension will be null
        templatecolumns = registerscales.DPmScales().getscalecolumns(self.name)
        columns = responsedict.items()
        count = {
            'extraversion': len(self.ocean['extraversion']),
            'emotional_stability': len(self.ocean['emotional_stability']),
            'conscientiousness': len(self.ocean['conscientiousness']),
            'openness': len(self.ocean['openness']),
            'agreeableness': len(self.ocean['agreeableness'])
        }
        check = {
            'extraversion': 0,
            'emotional_stability': 0,
            'conscientiousness': 0,
            'openness': 0,
            'agreeableness': 0
        }
        for colname in templatecolumns:
            for colname2,value in columns:
                if colname == colname2:
                    oceankey = re.search('^ipip_fifty_([0-9]+)',colname).groups()[0]
                    # check which dimension the key belongs in:
                    for dimension,d in self.ocean.items():
                        if oceankey in d:
                            if value != '': 
                                check[dimension] += 1

        # i need to set each dimension to null
        for dimension,chk in check.items():
            # compare response's answered question key count to normal question key count
            if chk != count[dimension]:
                self.score[dimension] = chk == 0 and registerscales.DPmNullFlags['empty'] or registerscales.DPmNullFlags['incomplete']
        
        # Calcuation:
        for colname in templatecolumns:
            for colname2,value in columns:
                if colname != colname2: continue
                if value == '' or value.isspace() : continue
                # Add calculation here:
                oceankey = re.search('^ipip_fifty_([0-9]+)',colname).groups()[0]

                # Total up scores for each dimension
                for dimension,dic in self.ocean.items():
                    # Skip if null
                    if self.score[dimension] in registerscales.DPmNullFlags.values(): continue 

                    if oceankey in dic:
                        if dic[oceankey] == 'positive':    
                            self.score[dimension] += self.positivekeys[value.lower()]
                        elif dic[oceankey] == 'negative': 
                            self.score[dimension] += self.negativekeys[value.lower()]
        return self        

    def report(self):
        print('IPIP Report:')
        for dimension,score in self.score.items():
            rangestr = 'Min: ' + str(self.dimensionranges[dimension]['negative']['total']) + ', Max: ' + str(self.dimensionranges[dimension]['positive']['total']) + ', Mdn: ' + str(self.dimensionranges[dimension]['median'])
            print(' '+dimension,':',score,'['+rangestr+']')

    def exportdata(self):
        return self.score

    def setprofile(self,pid):
        self.score = registerparticipants.DPmProfiles().getprofile(pid).getscaledata(self.name)

    def createtables(self,conn):
        conn.cursor().executescript('''
            DROP TABLE IF EXISTS IPIP50;

            CREATE TABLE IPIP50 (
                pid  TEXT NOT NULL PRIMARY KEY UNIQUE,
                score_openness INTEGER,
                score_conscientiousness INTEGER,
                score_extraversion INTEGER,
                score_agreeableness INTEGER,
                score_emotional_stability INTEGER
             );''')

    def exporttodb(self,conn,pid):
        self.setprofile(pid)

        conn.cursor().execute('''INSERT OR REPLACE INTO 
            IPIP50 (pid, score_openness, score_conscientiousness, score_extraversion, score_agreeableness, score_emotional_stability) 
            VALUES (?, ?, ?, ?, ?, ?)''',
         (pid, self.score['openness'],self.score['conscientiousness'],self.score['extraversion'],self.score['agreeableness'],self.score['emotional_stability']))
        conn.commit()

    def preparecsv(self,pid,IncludeHeader=True):
        self.setprofile(pid)

        headers = ['pid','ipip50_openness','ipip50_conscientiousness','ipip50_extraversion','ipip50_agreeableness','ipip50_emotional_stability']
        rows = [
            [pid,self.score['openness'],self.score['conscientiousness'],self.score['extraversion'],self.score['agreeableness'],self.score['emotional_stability']]
        ]

        return headers,rows

registerscales.DPmScales().registerscale(IPIP50())
